File: ProtocJsGenPlugin/TypeResolver.cs
using Google.Protobuf.Reflection;
using System.Collections.Generic;
using System.Linq;

namespace ProtocJsGenPlugin;

internal class TypeResolver
{
    // 类型全名 -> (proto文件名, 类型简单名)
    private readonly Dictionary<string, (string protoFile, string simpleName)> _typeMap = new();
    // 当前处理的proto文件
    private readonly FileDescriptorProto _currentFile;
    // 所有proto文件列表
    private readonly IReadOnlyList<FileDescriptorProto> _allProtoFiles;

    public TypeResolver(FileDescriptorProto currentFile, IReadOnlyList<FileDescriptorProto> allProtoFiles)
    {
        _currentFile = currentFile;
        _allProtoFiles = allProtoFiles;
        BuildTypeMap();
    }

    private void BuildTypeMap()
    {
        foreach (var protoFile in _allProtoFiles)
        {
            // 处理消息类型
            foreach (var message in protoFile.MessageType)
            {
                RegisterMessage(message, protoFile);
            }
            // 处理枚举类型
            foreach (var enumType in protoFile.EnumType)
            {
                RegisterEnum(enumType, protoFile);
            }
        }
    }

    private void RegisterMessage(DescriptorProto message, FileDescriptorProto protoFile, string parentFullName = "")
    {
        var fullName = GetFullName(message.Name, protoFile.Package, parentFullName);
        var simpleName = message.Name;
        _typeMap[fullName] = (protoFile.Name, simpleName);

        // 递归处理嵌套消息
        foreach (var nestedMessage in message.NestedType)
        {
            RegisterMessage(nestedMessage, protoFile, fullName);
        }

        // 处理嵌套枚举
        foreach (var nestedEnum in message.EnumType)
        {
            RegisterEnum(nestedEnum, protoFile, fullName);
        }
    }

    private void RegisterEnum(EnumDescriptorProto enumType, FileDescriptorProto protoFile, string parentFullName = "")
    {
        var fullName = GetFullName(enumType.Name, protoFile.Package, parentFullName);
        var simpleName = enumType.Name;
        _typeMap[fullName] = (protoFile.Name, simpleName);
    }

    private static string GetFullName(string name, string package, string parentFullName)
    {
        // 构建完整的类型名（以点开头）
        if (!string.IsNullOrEmpty(parentFullName))
        {
            return parentFullName + "." + name;
        }
        else if (!string.IsNullOrEmpty(package))
        {
            return "." + package + "." + name;
        }
        else
        {
            return "." + name;
        }
    }

    /// <summary>
    /// 获取类型信息，如果是外部类型则返回导入信息
    /// </summary>
    /// <param name="typeName">完整类型名（可能以点开头）</param>
    /// <returns>如果是外部类型，返回(proto文件名, 类型简单名)；否则返回null</returns>
    public (string protoFile, string simpleName)? GetExternalTypeInfo(string typeName)
    {
        // 规范化类型名：确保以点开头
        if (!typeName.StartsWith("."))
            typeName = "." + typeName;

        if (_typeMap.TryGetValue(typeName, out var info))
        {
            // 检查是否来自当前文件
            if (info.protoFile == _currentFile.Name)
                return null;

            return info;
        }

        // 未找到类型，可能是基础类型或错误
        return null;
    }

    /// <summary>
    /// 获取需要导入的类型列表（去重）
    /// </summary>
    /// <param name="referencedTypeNames">引用的类型名列表</param>
    /// <returns>需要导入的(proto文件名, 类型简单名)列表</returns>
    public List<(string protoFile, string simpleName)> GetRequiredImports(IEnumerable<string> referencedTypeNames)
    {
        var imports = new HashSet<(string protoFile, string simpleName)>();
        foreach (var typeName in referencedTypeNames)
        {
            var info = GetExternalTypeInfo(typeName);
            if (info.HasValue)
            {
                imports.Add(info.Value);
            }
        }
        return imports.ToList();
    }
}

File: src/type_resolver.cc
#include "type_resolver.h"

#include <algorithm>
#include <memory>
#include <string>
#include <unordered_set>
#include <utility>
#include <vector>

namespace protoc_js_gen_plugin {

namespace {

using google::protobuf::DescriptorProto;
using google::protobuf::EnumDescriptorProto;
using google::protobuf::FileDescriptorProto;

// Hash function for std::pair<std::string, std::string>
struct PairHash {
    std::size_t operator()(const std::pair<std::string, std::string>& p) const noexcept {
        std::size_t h1 = std::hash<std::string>{}(p.first);
        std::size_t h2 = std::hash<std::string>{}(p.second);
        // Combine hashes (from boost::hash_combine)
        return h1 ^ (h2 + 0x9e3779b9 + (h1 << 6) + (h1 >> 2));
    }
};

// Equality function for std::pair<std::string, std::string>
struct PairEqual {
    bool operator()(const std::pair<std::string, std::string>& a,
                    const std::pair<std::string, std::string>& b) const noexcept {
        return a.first == b.first && a.second == b.second;
    }
};

}  // namespace

TypeResolver::TypeResolver(
    const FileDescriptorProto& current_file,
    const std::vector<const FileDescriptorProto*>& all_proto_files)
    : current_file_(current_file),
      all_proto_files_(all_proto_files) {
    BuildTypeMap();
}

void TypeResolver::BuildTypeMap() {
    for (const FileDescriptorProto* proto_file : all_proto_files_) {
        // Process message types
        for (const DescriptorProto& message : proto_file->message_type()) {
            RegisterMessage(message, *proto_file);
        }

        // Process enum types
        for (const EnumDescriptorProto& enum_type : proto_file->enum_type()) {
            RegisterEnum(enum_type, *proto_file);
        }
    }
}

void TypeResolver::RegisterMessage(
    const DescriptorProto& message,
    const FileDescriptorProto& proto_file,
    const std::string& parent_full_name) {

    std::string full_name = GetFullName(message.name(), proto_file.package(), parent_full_name);
    type_map_[full_name] = std::make_pair(proto_file.name(), message.name());

    // Recursively process nested messages
    for (const DescriptorProto& nested_message : message.nested_type()) {
        RegisterMessage(nested_message, proto_file, full_name);
    }

    // Process nested enums
    for (const EnumDescriptorProto& nested_enum : message.enum_type()) {
        RegisterEnum(nested_enum, proto_file, full_name);
    }
}

void TypeResolver::RegisterEnum(
    const EnumDescriptorProto& enum_type,
    const FileDescriptorProto& proto_file,
    const std::string& parent_full_name) {

    std::string full_name = GetFullName(enum_type.name(), proto_file.package(), parent_full_name);
    type_map_[full_name] = std::make_pair(proto_file.name(), enum_type.name());
}

std::string TypeResolver::GetFullName(
    const std::string& name,
    const std::string& package,
    const std::string& parent_full_name) {

    // Build full type name (starting with dot)
    if (!parent_full_name.empty()) {
        return parent_full_name + "." + name;
    } else if (!package.empty()) {
        return "." + package + "." + name;
    } else {
        return "." + name;
    }
}

std::unique_ptr<TypeResolver::TypeInfo> TypeResolver::GetExternalTypeInfo(
    const std::string& type_name) const {

    // Normalize type name: ensure it starts with dot
    std::string normalized_name = type_name;
    if (normalized_name.empty() || normalized_name[0] != '.') {
        normalized_name = "." + normalized_name;
    }

    auto it = type_map_.find(normalized_name);
    if (it != type_map_.end()) {
        const TypeInfo& info = it->second;

        // Check if it's from the current file
        if (info.first == current_file_.name()) {
            return nullptr;
        }

        return std::make_unique<TypeInfo>(info);
    }

    // Type not found (could be a basic type or error)
    return nullptr;
}

std::vector<TypeResolver::TypeInfo> TypeResolver::GetRequiredImports(
    const std::vector<std::string>& referenced_type_names) const {

    std::unordered_set<TypeInfo, PairHash, PairEqual> imports;

    for (const std::string& type_name : referenced_type_names) {
        auto info = GetExternalTypeInfo(type_name);
        if (info) {
            imports.insert(*info);
        }
    }

    return std::vector<TypeInfo>(imports.begin(), imports.end());
}

}  // namespace protoc_js_gen_plugin